### PR TITLE
fix: handle container no image name provided [CAP-228]

### PR DESCRIPTION
**Pull request by iuliams**
#### Missed Anything?
- [x] Tests updated/written
- [x] merge only after snyk-docker-plugin is merged

#### What does this PR do?
This PR does the following:
* checks if the `container test` and `container monitor` commands receive an argument (the container image name) that shall be analysed;
* handles the errors returned by the scanning functions of the test and monitor commands

#### How should this be manually tested?
Old:
```
$ snyk container test
{"type":"Buffer","data":[]}
```
New:
```
$ snyk container test
Could not detect an image. Specify an image name to scan and try running the command again.
```
Same for the `monitor` command.

### CODE REVIEW

File: src/cli/commands/test/index.ts [hunk 1]
```diff
@@ -265,6 +266,11 @@ async function test(...args: MethodArgs): Promise<TestCommandResult> {
     error.code = errorResults[0].code;
     error.userMessage = errorResults[0].userMessage;
     error.strCode = errorResults[0].strCode;
+    if (options.docker) {
```
Comment (iuliams) [inline src/cli/commands/test/index.ts:269]: Ok, this is meant to fix the following corner case:
$ snyk container test /image
Since 'container test' not handled as the other commands that use the `ecosystem` [library](https://github.com/snyk/snyk/tree/master/src/lib/ecosystems), I decided to approach it in this manner as I'm not sure how big the change is otherwise on the 'docker-plugin' side. @RotemS please let me know your thoughts and if you see a nicer way to do this.

File: src/cli/commands/test/index.ts [hunk 2]
```diff
@@ -29,6 +29,7 @@ import {
   summariseErrorResults,
```
Comment (iuliams) [inline src/cli/commands/test/index.ts:29]: Added this as a separate commit as it may have to be re-iterated.

File: src/lib/ecosystems/monitor.ts [hunk 3]
```diff
@@ -52,8 +53,7 @@ export async function monitorEcosystem(
       ) {
         throw new DockerImageNotFoundError(path);
       }
-
-      throw error;
+      throw new InternalServerError("Scanning failed. Please make sure you are using the right parameters. If the issue still persists, please contact snyk support.");
```
Comment (iuliams) [inline src/lib/ecosystems/monitor.ts]: this is meant to fix e.g the following corner case (image name starting with '/'):
$ snyk container monitor /image

File: src/cli/commands/test/index.ts (hunk 1 above)
Comment (iuliams) [reply src/cli/commands/test/index.ts:269]: Hm actually, I just noticed that by simply using the ecosystem lib with the inner checks, this seems to scan the image correctly...:
```
--- a/src/lib/ecosystems/index.ts
+++ b/src/lib/ecosystems/index.ts
@@ -20,6 +20,9 @@ export function getEcosystemForTest(options: Options): Ecosystem | null {
   if (options.code) {
     return 'code';
   }
+  if (options.docker) {
+    return 'docker';
+  }
   return null;
 }
```

File: src/lib/errors/missing-arg-error.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import { CustomError } from './custom-error';
+
+export class MissingArgError extends CustomError {
+  constructor(command: string) {
+    const msg = `Could not detect an image to ${command}. Specify an image name to scan and try running the command again.`;
```
Comment (jahed-snyk) [inline src/lib/errors/missing-arg-error.ts:5]: Not blocking on this one:

It looks like we only need `method`/`command` to generate this message. Do we really need `method`? I think given the context of what the user ran, we can just say:

```
$ snyk test --docker
Could not detect an image. Specify an image name to scan and try running the command again.
```

This also avoids having to change method signatures for `processCommandArgs` and expose the concept of a "method" to it.

**Review by ghost — DISMISSED** [review]
Need to update tests but other than that looks good.

File: src/lib/ecosystems/monitor.ts [hunk 5]
```diff
@@ -52,8 +53,11 @@ export async function monitorEcosystem(
       ) {
         throw new DockerImageNotFoundError(path);
       }
-
-      throw error;
+      if (ecosystem === 'docker') {
+        throw new InternalServerError(
+          'Scanning failed. Please make sure you are using the right parameters. If the issue still persists, please contact snyk support.',
+        );
+      }
```
Comment (jahed-snyk) [inline src/lib/ecosystems/monitor.ts:60]: Do we have test coverage for this? We did we stop throwing `error`?

File: src/cli/commands/process-command-args.ts [hunk 6]
```diff
@@ -10,9 +10,11 @@ export function processCommandArgs<CommandOptions>(
...
   args = args.filter(Boolean);
 
-  // populate with default path (cwd) if no path given
   if (args.length === 0) {
-    args.unshift(process.cwd());
+    // For repository scanning, populate with default path (cwd) if no path given
+    if (!options.docker) {
+      args.unshift(process.cwd());
+    }
   }
```
Comment (jahed-snyk) [inline src/cli/commands/process-command-args.ts:18]: Should we combine into a single `if` statement?

File: src/lib/snyk-test/assemble-payloads.ts [hunk 7]
```diff
@@ -70,6 +71,17 @@ export async function assembleEcosystemPayloads(
     }
 
     return payloads;
+  } catch (error) {
+    if (ecosystem === 'docker' && error.message === 'authentication required') {
+      throw new DockerImageNotFoundError(options.path);
+    } else if (
```
Comment (jahed-snyk) [inline src/lib/snyk-test/assemble-payloads.ts:77]: Don't need the `else` here as previous statement is throwing. Might make it more readable as independant `if` statements.

File: src/cli/commands/monitor/index.ts [hunk 8]
```diff
@@ -75,6 +79,10 @@ async function monitor(...args0: MethodArgs): Promise<any> {
     throw new Error('`--remote-repo-url` is not supported for container scans');
   }
```
Comment (iuliams) [inline src/cli/commands/monitor/index.ts:81]: Add comment why we have to do this.

File: src/lib/snyk-test/assemble-payloads.ts (hunk 7 above)
Comment (iuliams) [inline src/lib/snyk-test/assemble-payloads.ts:77]: -> if

File: test/process-command-args.spec.ts [hunk 9]
```diff
@@ -9,6 +9,15 @@ describe('display help message', () => {
     expect(paths).toEqual([defaultCurrentPath]);
   });
 
+  it('should return expected options & paths when path is undefined and is a container run', () => {
```
Comment (jahed-snyk) [inline test/process-command-args.spec.ts:12]: Would be nice if these test cases were phrased with the expected result rather than just saying "expected" e.g. "should return no paths when running a container scan with no path"

File: src/cli/commands/test/index.ts [hunk 10]
```diff
@@ -68,6 +69,10 @@ async function test(...args: MethodArgs): Promise<TestCommandResult> {
   validateTestOptions(options);
   validateCredentials(options);
 
+  if (options.docker && paths.length === 0) {
+    throw new MissingArgError();
+  }
```
Comment (jahed-snyk) [inline src/cli/commands/test/index.ts:74]: It looks like this `test` function is trying to be Ecosystem agnostic (other than IAC) as part of Project Envelope IIRC. Should this statement be in `snyk-docker-plugin` (which is executed via `testEcosystem`)?

**Review by ghost — COMMENTED** [review]
I've left some comments. None of them are blocking this except the comment about Project Envelope. I don't have enough context on that so I'll see if someone else from the team can review my review.

File: src/cli/commands/test/index.ts (hunk 10 above)
Comment (iuliams) [reply src/cli/commands/test/index.ts:76]: Yes, totally agree and tried to make my way around it, but unfortunately for this specific check I cannot see another way because the docker plugin scan() is not even called, since there is no 'path' argument that it has to scan.
Let me know if you find the approach and added comment explanation ok for this scenario.

Comment (iuliams) [reply src/cli/commands/test/index.ts:76]: See e.g for monitoring:
https://github.com/snyk/snyk/blob/master/src/lib/ecosystems/monitor.ts#L41
Path is empty -> no plugin scan() -> with the current implementation it cannot be done in snyk-docker-plugin.
One idea would be to implement an extra validation plugin handler, but this would probably be another issue.

**Review by ghost — APPROVED** [review]

File: src/cli/commands/test/index.ts (hunk 10 above)
Comment (jahed-snyk) [reply src/cli/commands/test/index.ts:76]: That makes sense. I've approved this PR 👍

I agree, seems we're missing an ecosystem validation step or we're making too many assumptions around the concept of "paths". Something to improve on in the future.

File: src/lib/ecosystems/monitor.ts [hunk 11]
```diff
@@ -52,6 +52,9 @@ export async function monitorEcosystem(
       ) {
         throw new DockerImageNotFoundError(path);
       }
+      if (ecosystem === 'docker' && error.message === 'invalid image format') {
+        throw new DockerImageNotFoundError(path);
```
Comment (gitphill) [inline src/lib/ecosystems/monitor.ts:56]: looks same as exception thrown above - make one predicate that tests whether this exception should be thrown?

File: src/lib/snyk-test/assemble-payloads.ts [hunk 12]
```diff
@@ -70,6 +71,15 @@ export async function assembleEcosystemPayloads(
     }
 
     return payloads;
+  } catch (error) {
+    if (ecosystem === 'docker' && error.message === 'authentication required') {
+      throw new DockerImageNotFoundError(options.path);
+    }
+    if (ecosystem === 'docker' && error.message === 'invalid image format') {
+      throw new DockerImageNotFoundError(options.path);
+    }
```
Comment (gitphill) [inline src/lib/snyk-test/assemble-payloads.ts:80]: same as https://github.com/snyk/snyk/pull/2042/files#r662954528

**Review by gitphill — APPROVED** [review]

**Review by agranado2k — COMMENTED** [review]
I think it's missing the `snyk-docker-plugin` version being bumped.

File: src/lib/errors/missing-arg-error.ts [hunk 13]
```diff
@@ -0,0 +1,10 @@
+import { CustomError } from './custom-error';
+
+export class MissingArgError extends CustomError {
+  constructor() {
+    const msg = `Could not detect an image. Specify an image name to scan and try running the command again.`;
```
Comment (michael-c-snyk) [inline src/lib/errors/missing-arg-error.ts:5]: Is the use of ``` deliberate here? I guess the linter is happy /shrug

File: src/lib/ecosystems/monitor.ts (hunk 11 above)
Comment (iuliams) [reply src/lib/ecosystems/monitor.ts:56]: Thought about it, but then we discussed it's a bit more readable as such (without combining 4 conditions in one check). Wdyt?

**Review by gitphill — APPROVED** [review]

**Review by orsagie — APPROVED** [review]
